File: level1/lib/sensors/pcm_window_accumulator.dart
import 'dart:typed_data';

import '../core/models.dart';

/// Accumulates raw 16 kHz s16le PCM byte chunks (as produced by
/// `ffmpeg ... -ar 16000 -ac 1 -sample_fmt s16le -f s16le pipe:1`) into
/// [AudioWindow]s, converting to `Float32` samples (`sample / 32768`) and
/// carrying any partial window across chunk boundaries.
class PcmWindowAccumulator {
  final int sampleRate;
  final int windowSamples;
  final BytesBuilder _pending = BytesBuilder();

  PcmWindowAccumulator({
    this.sampleRate = 16000,
    this.windowSamples = 15600,
  });

  List<AudioWindow> add(Uint8List chunk) {
    _pending.add(chunk);
    final bytes = _pending.takeBytes();
    final sampleCount = bytes.length ~/ 2;
    final windows = <AudioWindow>[];
    var offset = 0;
    while (sampleCount - offset >= windowSamples) {
      windows.add(AudioWindow(
        timestamp: DateTime.now(),
        samples: _decode(bytes, offset, windowSamples),
        sampleRate: sampleRate,
      ));
      offset += windowSamples;
    }
    final consumedBytes = offset * 2;
    if (consumedBytes < bytes.length) {
      _pending.add(bytes.sublist(consumedBytes));
    }
    return windows;
  }

  Float32List _decode(Uint8List bytes, int sampleOffset, int count) {
    final out = Float32List(count);
    var p = sampleOffset * 2;
    final byteData = bytes.buffer.asByteData();
    for (var i = 0; i < count; i++) {
      out[i] = byteData.getInt16(p, Endian.little) / 32768;
      p += 2;
    }
    return out;
  }

  int get bufferedBytes => _pending.length;
}

File: level1/android/app/src/main/kotlin/io/securitycam/level1/sensors/PcmWindowAccumulator.kt
package io.securitycam.level1.sensors

import io.securitycam.level1.detection.AudioWindow
import java.time.Instant

/**
 * Accumulates raw 16 kHz s16le PCM byte chunks into [AudioWindow]s, converting
 * to Float32 samples (`sample / 32768`) and carrying any partial window across
 * chunk boundaries (port of `lib/sensors/pcm_window_accumulator.dart`).
 */
class PcmWindowAccumulator(
    val sampleRate: Int = 16000,
    val windowSamples: Int = 15600,
) {
    private val pending = ArrayDeque<Byte>()

    fun add(chunk: ByteArray): List<AudioWindow> {
        for (b in chunk) pending.add(b)
        val bytes = ByteArray(pending.size)
        var i = 0
        for (b in pending) bytes[i++] = b
        pending.clear()
        val windows = mutableListOf<AudioWindow>()
        var sampleOffset = 0
        while (bytes.size / 2 - sampleOffset >= windowSamples) {
            windows.add(
                AudioWindow(
                    timestamp = Instant.now(),
                    samples = decode(bytes, sampleOffset, windowSamples),
                    sampleRate = sampleRate,
                ),
            )
            sampleOffset += windowSamples
        }
        val consumedBytes = sampleOffset * 2
        if (consumedBytes < bytes.size) {
            for (b in bytes.copyOfRange(consumedBytes, bytes.size)) pending.add(b)
        }
        return windows
    }

    private fun decode(bytes: ByteArray, sampleOffset: Int, count: Int): FloatArray {
        val out = FloatArray(count)
        var p = sampleOffset * 2
        for (j in 0 until count) {
            val lo = bytes[p].toInt() and 0xFF
            val hi = bytes[p + 1].toInt() and 0xFF
            val v = (lo or (hi shl 8)).toShort().toInt()
            out[j] = v / 32768.0f
            p += 2
        }
        return out
    }

    val bufferedBytes: Int get() = pending.size
}